src/cljam/io/wig.clj

```diff
@@ -0,0 +1,171 @@
+  "Normalize WIG lines."
+  [m]
+  (update m :chrom chr/normalize-chromosome-key))
+
+(defn- fields->map
+  "Convert vector [\"key1=value1\" \"key2=value2\" ...] to map {:key1 \"value1\" :key2 \"value2\" ...}."
+  [fields]
+  (->> fields
+       (map #(cstr/split % #"="))
+       (map (fn [[k v]] [(keyword k) v]))
+       (into {})))
+
+(defn- str->wiggle-track-data
+  "Convert string to Wiggle track data values which can be integer or real,
+  positive or negative values."
+  [^String s]
+  (if-let [l (as-long s)]
+    l
+    (as-double s)))
+
+(defn- deserialize-wigs
+  "Parse WIG lines and returns a vector of maps."
+  [lines]
+  (loop [lines lines
+         wigs []
+         current-track nil]
+    (if (nil? (first lines))
+      (conj wigs current-track)
+      (let [fields (-> lines first cstr/trim (cstr/split #"\s+"))]
+        (case (first fields)
+          ; declaration line of variableStep
+          "variableStep"
+          (let [{:keys [chrom span] :or {span 1}} (->> fields rest fields->map)]
+            (recur (rest lines)
+                   (cond-> wigs
+                     (some? current-track) (conj current-track))
+                   {:format :variable-step
+                    :chrom chrom
+                    :span (as-long span)
+                    :values []}))
+
+          ; declaration line of fixedStep
+          "fixedStep"
+          (let [{:keys [chrom start step span] :or {span 1}} (->> fields rest fields->map)]
+            (recur (rest lines)
+                   (cond-> wigs
+                     (some? current-track) (conj current-track))
+                   {:format :fixed-step
+                    :chrom chrom
+                    :start (as-long start)
+                    :step (as-long step)
+                    :span (as-long span)
+                    :values []}))
+
+          ; data line
+          (let [v (case (:format current-track)
+                    :variable-step
+                    (let [[start value] fields]
+                      {:start (as-long start)
+                       :value (str->wiggle-track-data value)})
+
+                    :fixed-step
+                    (-> fields first str->wiggle-track-data)
+
+                    (throw (IllegalArgumentException. "Invalid wiggle format")))]
+            (recur (rest lines)
+                   wigs
+                   (update current-track :values #(conj % v)))))))))
+
+(defn- serialize-wig
+  "Serialize WIG line into vector of strings."
+  [wig]
+  {:pre [;; There are two options for formatting wiggle data.
+         (or (= (:format wig) :fixed-step) (= (:format wig) :variable-step))
+         ;; These fields are required if the format is fixedStep.
+         (if (= (:format wig) :fixed-step) (and (:chrom wig) (:span wig) (:start wig) (:step wig) (:values wig)) true)
+         ;; These fields are required if the format is variableStep.
+         (if (= (:format wig) :variable-step) (and (:chrom wig) (:span wig) (:values wig)) true)]}
```

Comment:
These preconditions are somewhat too complicated. I'd take them away from the function and define a new dedicated predicate to validate the argument WIG data.

Also, since the existence of `span` and `step` parameters are not mandatory (as mentioned in some other comments), drop those conditions from here.

Reply:
Ah, let me rephrase that. If you feed default values for those (`span` and `step`) parameters, there's no need to change the conditions at all.